File: app/src/main/java/com/jtrofe/cheesebots/physics/Vec.java
package com.jtrofe.cheesebots.physics;

import java.util.List;
import java.util.Random;

/**
 * A 2-dimensional vector class
 */
public class Vec {

    public double x;
    public double y;

    public int xi(){
        return (int) x;
    }
    public int yi(){
        return (int) y;
    }

    public float xf(){
        return (float) x;
    }
    public float yf(){
        return (float) y;
    }

    //
    //  Static methods
    //
    public static Vec Random(Vec max){
        Random r = new Random();

        return new Vec(r.nextDouble() * max.x, r.nextDouble() * max.y);
    }

    public static Vec RandomDir(double length){
        Random rnd = new Random();
        double l = length * 0.75 * rnd.nextDouble() + (length * 0.5);
        double angle = Math.PI * 2 * rnd.nextDouble();

        double dx = Math.sin(angle);
        double dy = -Math.cos(angle);

        return new Vec(dx * l, dy * l);
    }

    /**
     * Find the point in a list which is closest to a goal point
     * @param pointList A list of points
     * @param goal The goal point
     * @return Closest point
     */
    public static Vec GetClosestToPoint(List<Vec> pointList, Vec goal){
        double closest_distance = Double.MAX_VALUE;
        Vec closest_point = pointList.get(0);

        for(Vec v:pointList){
            double d = v.Subtract(goal).LengthSquared();

            if(d < closest_distance){
                closest_distance = d;
                closest_point = v.copy();
            }
        }

        return closest_point;
    }

    //
    //  Constructors
    //
    public Vec(){
        this.x = 0;
        this.y = 0;
    }

    public Vec(double x, double y){
        this.x = x;
        this.y = y;
    }

    public Vec copy(){
        return new Vec(x, y);
    }

    //
    //  Math operations
    //
    public Vec Subtract(Vec v){
        return new Vec(x - v.x, y - v.y);
    }

    public Vec Add(Vec v){
        return new Vec(x + v.x, y + v.y);
    }

    public Vec ScalarMultiply(double s){
        return new Vec(x * s, y * s);
    }

    public Vec ScalarDivide(double s){
        return new Vec(x / s, y / s);
    }

    public double Dot(Vec v){
        return x * v.x + y * v.y;
    }

    public Vec Normalize(){
        if(x == 0 && y == 0) return new Vec();

        return this.ScalarDivide(this.Length());
    }

    public Vec Clamp(double maxLength){
        if(this.LengthSquared() > (maxLength * maxLength)){
            return this.Normalize().ScalarMultiply(maxLength);
        }

        return new Vec(x, y);
    }

    public Vec Negate(){
        return new Vec(-x, -y);
    }

    public double LengthSquared(){
        return x * x + y * y;
    }

    public double Length(){
        return Math.sqrt(x * x + y * y);
    }

    @Override
    public String toString(){
        return "{" + x + ", " + y + "}";
    }
}

File: app/src/main/java/com/jtrofe/cheesebots/physics/Vec.kt
package com.jtrofe.cheesebots.physics

import android.graphics.PointF
import java.util.ArrayList
import java.util.Random

/**
 * Created by MAIN on 2/8/16.
 */
public data class Vec(var x:Double = 0.0, var y:Double = 0.0){

    var xf:Float = x.toFloat()
        private set
        get() = x.toFloat()

    var yf:Float = y.toFloat()
        private set
        get() = y.toFloat()

    var xi:Int = x.toInt()
        private set
        get() = x.toInt()

    var yi:Int = y.toInt()
        private set
        get() = y.toInt()


    companion object{
        public fun Random(max:Vec):Vec{
            val r:Random = Random()

            return Vec(max.x * r.nextDouble(), max.y * r.nextDouble())
        }

        /**
         * Find the point in a list which is closest to a goal point
         * @param pointList A list of points
         * @param goal The goal point
         * @return Closest point
         */
        public fun GetClosestToPoint(pointList:List<Vec>, goal:Vec):Vec{
            var closest_distance = Double.MAX_VALUE
            var closest_point = pointList[0]

            pointList.forEach{
                val d = (it - goal).LengthSquared()

                if(d < closest_distance){
                    closest_distance = d
                    closest_point = it.copy()
                }
            }

            return closest_point
        }

        public fun GetVectorsFromPoint(pointList:List<Vec>, start:Vec):List<Vec>{
            return pointList.map{ it -> it - start}
        }

        /*public static Vec[] GetVectorsFromPoint(Vec[] pointList, Vec start){
        Vec[] returnList = new Vec[pointList.length];

        for(int i=0;i<pointList.length;i++){
            returnList[i] = pointList[i].Subtract(start);
        }

        return returnList;
    }*/
    }

    constructor(point:PointF): this(point.x.toDouble(), point.y.toDouble()){}
    constructor(x:Int, y:Int): this(x.toDouble(), y.toDouble()){}
    constructor(x:Float, y:Float): this(x.toDouble(), y.toDouble()){}

    //
    //  Operator overloads
    //
    public fun plus(v:Vec):Vec{
        return Vec(x + v.x, y + v.y)
    }

    public fun minus(v:Vec):Vec{
        return Vec(x - v.x, y - v.y)
    }

    public fun times(s:Double):Vec{
        return Vec(x * s, y * s)
    }

    public fun times(s:Int):Vec{
        return Vec(x * s.toDouble(), y * s.toDouble())
    }

    public fun div(s:Double):Vec{
        return Vec(x / s, y / s)
    }

    public fun div(s:Int):Vec{
        return Vec(x / s.toDouble(), y / s.toDouble())
    }

    public fun minus():Vec{
        return Vec(-x, -y)
    }

    //
    //  Calculations
    //
    public fun LengthSquared():Double{
        return x * x + y * y
    }

    public fun Length():Double{
        return Math.sqrt(LengthSquared())
    }

    //
    //  Other operations
    //
    public fun Dot(v2:Vec):Double{
        return x * v2.x + y * v2.y
    }

    public fun Normalize():Vec{
        if(x.equals(0.0) && y.equals(0.0)) return Vec()

        val length = Length()
        return this / length
    }

    public fun Clamp(maxLength:Double):Vec{
        if(LengthSquared() > (maxLength * maxLength)){
            return Normalize() * maxLength
        }

        return Vec(x, y)
    }
}